os/drivers/audio/alc5658.c

```diff
@@ -974,8 +974,31 @@ static int alc5658_ioctl(FAR struct audio_lowerhalf_s *dev, int cmd, unsigned lo
 	switch (cmd) {
 
 	case AUDIOIOC_PREPARE: {
-		/* TODO: Need to be implemented later */
-		audvdbg("AUDIOIOC_PREPARE: No Action Would be taken now \n");
+		/* This is to aid audio prepare, this would stop dma, i2s and
+		cleanup all queues. Idea behind is to use this in xrun cases to stop
+		followed by (restart) but restart is not part of this */
+		audvdbg("AUDIOIOC_PREPARE: Stop dma,i2s and cleanup qs\n");
+
+		/* Take semaphore */
+		alc5658_takesem(&priv->devsem);
+
+		/* Stop dma, i2s */
+		I2S_STOP(priv->i2s);
+
+		/* Switch alc status */
+		priv->running = false;
+
+		/* Clean up alc queue */
+		sq_entry_t *tmp = NULL;
+		for (tmp = (sq_entry_t *)sq_peek(&priv->pendq); tmp; tmp = sq_next(tmp)) {
+			sq_rem(tmp, &priv->pendq);
+		}
+		sq_init(&priv->pendq);
+
+		/* TOCHECK: Do we need to run stop alc stop script? ??? for now no */
+
```

Comment:
In my opinion, pcm_prepare is being done to restart the pcm.. so either we have to call codec_stop or atleast we have to call codec_reset_script ... but need to make sure that codec_reset_script only resets the codec without altering the codec configurations.. like sample rate, etc..

Reply:
NO. Codec reset script would reset codec that means all registers will have default values. It has to be reconfigured again. Not a good idea. But, if what you are saying is right, then using pcm_prepare to handle xrun may be wRONG. Instead tinyalsa has to stop and start again. Correct?

Reply:
Please move codec init and configuration code from start() api to this place in "prepare".
After codec configuration, pls call I2S_RESUME.